pdata/pmetric/json.go

```diff
@@ -64,3 +68,603 @@ func (d *jsonUnmarshaler) UnmarshalMetrics(buf []byte) (Metrics, error) {
+	td := readMetricsData(iter)
+	err := iter.Error
+	if err != nil {
+		return Metrics{}, err
+	}
+	return internal.MetricsFromProto(td), err
+}
+
+func readMetricsData(iter *jsoniter.Iterator) otlpmetrics.MetricsData {
+	td := otlpmetrics.MetricsData{}
+	iter.ReadObjectCB(func(iter *jsoniter.Iterator, f string) bool {
+		switch f {
+		case "resource_metrics", "resourceMetrics":
+			iter.ReadArrayCB(func(iterator *jsoniter.Iterator) bool {
+				td.ResourceMetrics = append(td.ResourceMetrics, readResourceMetrics(iter))
+				return true
+			})
+		default:
+			iter.ReportError("readMetricsData", fmt.Sprintf("unknown field:%v", f))
+		}
+		return true
+	})
+	return td
+}
+
+func readResourceMetrics(iter *jsoniter.Iterator) *otlpmetrics.ResourceMetrics {
+	rs := &otlpmetrics.ResourceMetrics{}
+	iter.ReadObjectCB(func(iter *jsoniter.Iterator, f string) bool {
+		switch f {
+		case "resource":
+			iter.ReadObjectCB(func(iter *jsoniter.Iterator, f string) bool {
+				switch f {
+				case "attributes":
+					iter.ReadArrayCB(func(iter *jsoniter.Iterator) bool {
+						rs.Resource.Attributes = append(rs.Resource.Attributes, readAttribute(iter))
+						return true
+					})
+				case "droppedAttributesCount", "dropped_attributes_count":
+					rs.Resource.DroppedAttributesCount = iter.ReadUint32()
+				default:
+					iter.ReportError("readResourceMetrics.resource", fmt.Sprintf("unknown field:%v", f))
+				}
+				return true
+			})
+		case "instrumentationLibraryMetrics", "instrumentation_library_metrics", "scopeMetrics", "scope_metrics":
+			iter.ReadArrayCB(func(iter *jsoniter.Iterator) bool {
+				rs.ScopeMetrics = append(rs.ScopeMetrics,
+					readInstrumentationLibraryMetrics(iter))
+				return true
+			})
+		case "schemaUrl", "schema_url":
+			rs.SchemaUrl = iter.ReadString()
+		default:
+			iter.ReportError("readResourceMetrics", fmt.Sprintf("unknown field:%v", f))
+		}
+		return true
+	})
+	return rs
+}
+
+func readInstrumentationLibraryMetrics(iter *jsoniter.Iterator) *otlpmetrics.ScopeMetrics {
+	ils := &otlpmetrics.ScopeMetrics{}
+	iter.ReadObjectCB(func(iter *jsoniter.Iterator, f string) bool {
+		switch f {
+		case "instrumentationLibrary", "instrumentation_library", "scope":
+			iter.ReadObjectCB(func(iter *jsoniter.Iterator, f string) bool {
+				switch f {
+				case "name":
+					ils.Scope.Name = iter.ReadString()
+				case "version":
+					ils.Scope.Version = iter.ReadString()
+				default:
+					iter.ReportError("instrumentationLibrary", fmt.Sprintf("unknown field:%v", f))
+				}
+				return true
+			})
+		case "metrics":
+			iter.ReadArrayCB(func(iter *jsoniter.Iterator) bool {
+				ils.Metrics = append(ils.Metrics, readMetric(iter))
+				return true
+			})
+		case "schemaUrl", "schema_url":
+			ils.SchemaUrl = iter.ReadString()
+		default:
+			iter.ReportError("readInstrumentationLibraryMetrics", fmt.Sprintf("unknown field:%v", f))
+		}
+		return true
+	})
+	return ils
+}
+
+func readMetric(iter *jsoniter.Iterator) *otlpmetrics.Metric {
```

Comment:
I would split this into readSumMetric/readGaugeMetric/... Would be easier to review.